File: src/pcl2odomc.py
#!/usr/bin/env python

import rospy
import numpy
import math
import message_filters
import sensor_msgs.point_cloud2 as PCL
from sensor_msgs.msg import PointCloud2, PointField
from nav_msgs.msg import Odometry
import time

def pointcloud2_to_list(cloud):
    gen = PCL.read_points(cloud, skip_nans=True, field_names=('x', 'y', 'z'))
    list_of_tuples = list(gen)
    return list_of_tuples


def list_to_pointcloud2(points):
    pcloud = PointCloud2()
    pcloud = PCL.create_cloud_xyz32(pcloud.header, points)
    pcloud.header.stamp = rospy.Time.now()
    pcloud.header.frame_id = '/odom'
    return pcloud

new_scan = None
new_odomc = None
is_running = False
pub = None

def callback(odomc, scan):
    #rospy.loginfo('Delta stamp: %ds %dns' % (odomc.header.stamp.secs - scan.header.stamp.secs, odomc.header.stamp.nsecs - scan.header.stamp.nsecs))
    t = [odomc.pose.pose.position.x, odomc.pose.pose.position.y, odomc.pose.pose.position.z]
    q = [odomc.pose.pose.orientation.w, odomc.pose.pose.orientation.x,
         odomc.pose.pose.orientation.y, odomc.pose.pose.orientation.z]
    transformed_scan = transform_scan(pointcloud2_to_list(scan), t, q)
    pub.publish(list_to_pointcloud2(transformed_scan))

def compute_rotation_matrix(q):
    return numpy.matrix([
        [q[0] ** 2 + q[1] ** 2 - q[2] ** 2 - q[3] ** 2, 2 * (q[1] * q[2] - q[0] * q[3]),
         2 * (q[1] * q[3] + q[0] * q[2])],
        [2 * (q[1] * q[2] + q[0] * q[3]), q[0] ** 2 + q[2] ** 2 - q[1] ** 2 - q[3] ** 2,
         2 * (q[2] * q[3] - q[0] * q[1])],
        [2 * (q[1] * q[3] - q[0] * q[2]), 2 * (q[2] * q[3] + q[0] * q[1]),
         q[0] ** 2 + q[3] ** 2 - q[1] ** 2 - q[2] ** 2]
    ])

def transform_scan(scan, t, q):
    new_scan = []
    rotation_matrix = compute_rotation_matrix(q)
    translation_vector = numpy.matrix(t).getT()
    for i in range(len(scan)):
        new_scan.append(list((rotation_matrix * numpy.matrix(scan[i]).getT() + translation_vector).flat))
    return new_scan

def main():
    rospy.init_node('Pcl2odomc')
    odom_sub = message_filters.Subscriber('/odomc', Odometry)
    scan_sub = message_filters.Subscriber('/cloud_out', PointCloud2)
    global pub
    pub = rospy.Publisher('/icp/scan', PointCloud2, queue_size=10)
    ts = message_filters.ApproximateTimeSynchronizer([odom_sub, scan_sub], 50, 0.5)
    ts.registerCallback(callback)
    global new_scan
    global new_odomc
    global is_running
    while not rospy.is_shutdown():
        '''if new_scan is not None and new_odomc is not None:
            print('Running:  ' + str(is_running))
            is_running = True
            scan = new_scan
            odomc = new_odomc
            t = [odomc.pose.pose.position.x, odomc.pose.pose.position.y, odomc.pose.pose.position.z]
            q = [odomc.pose.pose.orientation.w, odomc.pose.pose.orientation.x,
                 odomc.pose.pose.orientation.y, odomc.pose.pose.orientation.z]
            transformed_scan = transform_scan(pointcloud2_to_list(scan), t, q)
            pub.publish(list_to_pointcloud2(transformed_scan))
            new_scan = None
            new_odomc = None
            is_running = False
            print('Completed: ' + str(is_running))'''
        pass
if __name__ == "__main__":
    main()

File: src/pcl2odomc.cpp
#include "ros/ros.h"
#include "tf/transform_listener.h"
#include "sensor_msgs/PointCloud2.h"
#include "tf/message_filter.h"
#include "message_filters/subscriber.h"
#include "laser_geometry/laser_geometry.h"
#include "pcl_ros/transforms.h"
#include "pcl/point_types.h"
#include "tf_conversions/tf_eigen.h"
#include "nav_msgs/Odometry.h"
#include "geometry_msgs/PoseStamped.h"
#include <string>
ros::Publisher tf_pub;

int main(int argc, char** argv)
{
	ros::init(argc, argv, "point_cloud_to_map");

	ros::NodeHandle n;
	ros::Duration(2).sleep();
	tf::TransformListener _tf;
    tf::Transformer tflooker;
    ros::Duration(2).sleep();
    tf::StampedTransform transform;
	while(true){
	nav_msgs::Odometry::ConstPtr msg = ros::topic::waitForMessage<nav_msgs::Odometry>(std::string("/odom"), ros::Duration(1.0));
	nav_msgs::Odometry::ConstPtr msg_odomc = ros::topic::waitForMessage<nav_msgs::Odometry>(std::string("/odomc"), ros::Duration(1.0));
	geometry_msgs::PoseStamped transformed, pose;

	pose.pose = msg->pose.pose;
	pose.header.stamp = msg->header.stamp;
	pose.header.frame_id = msg->header.frame_id;
	//tflooker.lookupTransform(std::string("/odom"), std::string("/map"), ros::Time(0), transform);
	try{
	ROS_INFO("Original Pose-> x: [%f], y: [%f], z: [%f]", pose.pose.position.x, pose.pose.position.y, pose.pose.position.z);
	ROS_INFO("Original Orientation-> x: [%f], y: [%f], z: [%f], w: [%f]", pose.pose.orientation.x, pose.pose.orientation.y, pose.pose.orientation.z, pose.pose.orientation.w);
	_tf.transformPose(std::string("/map"), pose, transformed);
	tf::StampedTransform trf;
	_tf.lookupTransform("odom", "map", ros::Time(0), trf);

	tf::Vector3 tr = trf.getOrigin();
	ROS_INFO("transformListener: T x: [%f], y: [%f], z: [%f]", tr[0], tr[1], tr[2]);
	tf::Quaternion q = trf.getRotation();
	ROS_INFO("transformListener: R x: [%f], y: [%f], z: [%f], w: [%f]", q[0], q[1], q[2], q[3]);

	ROS_INFO("Transformed Pose-> x: [%f], y: [%f], z: [%f]", transformed.pose.position.x, transformed.pose.position.y, transformed.pose.position.z);
	ROS_INFO("Transformed Orientation-> x: [%f], y: [%f], z: [%f], w: [%f]", transformed.pose.orientation.x, transformed.pose.orientation.y, transformed.pose.orientation.z, transformed.pose.orientation.w);
	}
	catch(...) {
	}
    }
	return 0;
}
